File: solutions/union-find/typescript/0990-satisfiability.ts
/**
### INTUITION:
This is a classic **Union-Find** problem. We need to check if equality and inequality constraints can be satisfied simultaneously. The key insight is to first process all equality constraints to group variables, then check if inequality constraints violate these groups.

### APPROACH:
1. **Initialize parent array**: Create Array (parent tracking for **Union-Find**) of size 26 for tracking equivalence classes (one per lowercase letter)
2. **Process equalities**: Use **Union-Find** to group variables that must be equal, storing parent relationships in array
3. **Check inequalities**: For each "!=" constraint using String (equation parsing), verify variables are in different groups using find operation on array
4. **Return result**: True if no conflicts found, False otherwise

### WHY THIS WORKS:
- **Union-Find** efficiently manages equivalence classes through array-based parent tracking
- Path compression optimizes find operations by flattening tree structure in array
- **Array** traversal processes equations in two passes for efficiency
- Equality constraints create connected components (equivalence classes)
- Inequality constraints must not connect variables in same component
- Two-pass approach separates grouping from validation

### EXAMPLE WALKTHROUGH:
Given input ["a==b","b==c","a!=d"]:

Input:
```
["a==b","b==c","a!=d"]
```

Step 1:** Initialize parent array
- parent[0..25] = [0,1,2,...,25] (each variable is its own parent)
- Variables: a=0, b=1, c=2, d=3

Step 2:** Process equality "a==b"
- Union(a, b): parent[0] = 1
- Groups: {a,b}, {c}, {d}, ...

Step 3:** Process equality "b==c"
- Union(b, c): parent[1] = 2
- Groups: {a,b,c}, {d}, ...

Step 4:** Check inequality "a!=d"
- find(a) = 2 (through path compression: a→b→c)
- find(d) = 3
- 2 ≠ 3, so no conflict

Step 4:** Return result
- All constraints satisfied, no contradictions found
- Can assign: a=0, b=0, c=0, d=1

Output:
```
true
```

### TIME COMPLEXITY:
O(N × α(N)**)
Where N is number of equations and α is inverse Ackermann function

### SPACE COMPLEXITY:
O(1)**
- Constant extra space
Since we only have 26 possible variables (a-z)

### EDGE CASES:
- **Empty input**: Handle when input is empty
- **Single element**: Handle single-element inputs
- **Boundary values**: Handle minimum/maximum valid values

*/

class Solution {
  /**
   * Check if equality equations can be satisfied using **Union-Find**.
   *
   *         Args:
   *             equations: List of equations in format "xi==xj" or "xi!=xj"
   *
   *         Returns:
   *             True if equations can be satisfied, False otherwise
   *
   *         Time Complexity: O(N × α(N)) where N is number of equations
   *         Space Complexity: O(1) since we have at most 26 variables
   */
  equationsPossible(equations: string[]): boolean {
    const parent = **Array**.from({ length: 26 }, (_, i) => i);

    const find = (x: number): number => {
      if (parent[x] !== x) {
        parent[x] = find(parent[x]); // Path compression
      }
      return parent[x];
    };

    const union = (x: number, y: number): void => {
      parent[find(x)] = find(y);
    };

    // Process all equality equations first
    for (const eq of equations) {
      if (eq[1] === "=") {
        const x = eq.charCodeAt(0) - 97; // 'a' = 97
        const y = eq.charCodeAt(3) - 97;
        union(x, y);
      }
    }

    // Check inequality equations
    for (const eq of equations) {
      if (eq[1] === "!") {
        const x = eq.charCodeAt(0) - 97;
        const y = eq.charCodeAt(3) - 97;
        if (find(x) === find(y)) {
          return false; // Contradiction found
        }
      }
    }

    return true;
  }

  /**
   * Alternative implementation with explicit **Union-Find** class.
   *
   *         Args:
   *             equations: List of equations
   *
   *         Returns:
   *             True if satisfiable, False otherwise
   */
  equationsPossibleAlternative(equations: string[]): boolean {
    class UnionFind {
      parent: number[];
      rank: number[];

      constructor(n: number) {
        this.parent = **Array**.from({ length: n }, (_, i) => i);
        this.rank = **Array**(n).fill(0);
      }

      find(x: number): number {
        if (this.parent[x] !== x) {
          this.parent[x] = this.find(this.parent[x]);
        }
        return this.parent[x];
      }

      union(x: number, y: number): void {
        const rootX = this.find(x);
        const rootY = this.find(y);

        if (rootX !== rootY) {
          if (this.rank[rootX] < this.rank[rootY]) {
            this.parent[rootX] = rootY;
          } else if (this.rank[rootX] > this.rank[rootY]) {
            this.parent[rootY] = rootX;
          } else {
            this.parent[rootY] = rootX;
            this.rank[rootX]++;
          }
        }
      }
    }

    const uf = new UnionFind(26);

    // Process equalities
    for (const eq of equations) {
      if (eq[1] === "=") {
        const x = eq.charCodeAt(0) - 97;
        const y = eq.charCodeAt(3) - 97;
        uf.union(x, y);
      }
    }

    // Check inequalities
    for (const eq of equations) {
      if (eq[1] === "!") {
        const x = eq.charCodeAt(0) - 97;
        const y = eq.charCodeAt(3) - 97;
        if (uf.find(x) === uf.find(y)) {
          return false;
        }
      }
    }

    return true;
  }
}

// Test cases
if (typeof module !== "undefined" && module.exports) {
  module.exports = Solution;
}

function runTests(): void {
  const solution = new Solution();

  const equations = ["a==b", "b==c", "a!=d"];
  console.log(`Equations [${equations}] satisfiable:`, solution.equationsPossible(equations));

  const equations2 = ["a==b", "b!=a"];
  console.log(`Equations [${equations2}] satisfiable:`, solution.equationsPossible(equations2));
}

if (typeof require !== "undefined" && require.main === module) {
  runTests();
}

export default Solution;

File: solutions/union-find/python/0990-satisfiability.py
"""
### INTUITION:
This is a classic Union-Find problem. We need to check if equality and inequality constraints can be satisfied simultaneously. The key insight is to first process all equality constraints to group variables, then check if inequality constraints violate these groups.

### APPROACH:
1. **Process equalities**: Use Union-Find to group variables that must be equal
2. **Check inequalities**: For each "!=" constraint, verify variables are in different groups
3. **Return result**: True if no conflicts found, False otherwise

### WHY THIS WORKS:
- Union-Find efficiently manages equivalence classes
- Equality constraints create connected components
- Inequality constraints must not connect variables in same component
- Two-pass approach separates grouping from validation

### EXAMPLE WALKTHROUGH:
Input:
```
["a==b","b!=a"]
["a==b","b==c","a!=d"]
```

Step 1: Process equalities - Union('a', 'b')
Step 2: Check inequalities - "b!=a" but a and b are in same group

Steps:
Step 1: Process equalities - Union('a','b'), Union('b','c') → {a,b,c} group
Step 2: Check inequalities - "a!=d" and d is separate → no conflict

Output:
```
False (contradiction)
True
```

### TIME COMPLEXITY:
**O(N × α(N)**)
Where N is number of equations and α is inverse Ackermann function

### SPACE COMPLEXITY:
**O(1)**
Since we only have 26 possible variables (a-z)

### EDGE CASES:
- **Empty input**: Handle when input is empty
- **Single element**: Handle single-element inputs
- **Boundary values**: Handle minimum/maximum valid values

</details>

"""

from typing import Any
import re


class UnionFind:
    """Union-Find (Disjoint Set Union) data structure."""

    def __init__(self, n: int) -> None:
        """Initialize with n elements."""
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x: int) -> int:
        """Find root of element x with path compression."""
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def union(self, x: int, y: int) -> bool:
        """Union two sets. Returns True if they were in different sets."""
        px, py = self.find(x), self.find(y)
        if px == py:
            return False

        if self.rank[px] < self.rank[py]:
            self.parent[px] = py
        elif self.rank[px] > self.rank[py]:
            self.parent[py] = px
        else:
            self.parent[py] = px
            self.rank[px] += 1

        return True

    def connected(self, x: int, y: int) -> bool:
        """Check if two elements are in the same set."""
        return self.find(x) == self.find(y)

    @property
    def components(self) -> int:
        """Return number of connected components."""
        return len(set(self.find(i) for i in range(len(self.parent))))


class Solution:
    def equationsPossible(self, equations: list[str]) -> bool:
        """
        Check if equality equations can be satisfied using Union-Find.

        Args:
            equations: List of equations in format "xi==xj" or "xi!=xj"

        Returns:
            True if equations can be satisfied, False otherwise

        Time Complexity: O(N × α(N)) where N is number of equations
        Space Complexity: O(1) since we have at most 26 variables
        """
        # Union-Find for 26 lowercase letters
        parent = list(range(26))

        def find(x: Any) -> Any:
            """Find root with path compression."""
            if parent[x] != x:
                parent[x] = find(parent[x])
            return parent[x]

        def union(x: Any, y: Any) -> Any:
            """Union two variables."""
            px, py = find(x), find(y)
            if px != py:
                parent[px] = py

        def char_to_index(c: Any) -> Any:
            """Convert character to index (a=0, b=1, etc.)."""
            return ord(c) - ord("a")

        # Step 1: Process all equality equations
        for eq in equations:
            if eq[1] == "=":  # Equality equation "xi==xj"
                x, y = char_to_index(eq[0]), char_to_index(eq[3])
                union(x, y)

        # Step 2: Check all inequality equations
        for eq in equations:
            if eq[1] == "!":  # Inequality equation "xi!=xj"
                x, y = char_to_index(eq[0]), char_to_index(eq[3])
                if find(x) == find(y):  # Same component = contradiction
                    return False

        return True

    def equationsPossibleAlternative(self, equations: list[str]) -> bool:
        """
        Alternative implementation with explicit Union-Find class.

        Args:
            equations: List of equations

        Returns:
            True if satisfiable, False otherwise
        """

        class UnionFind:
            def __init__(self: Any, n: Any) -> None:
                self.parent = list(range(n))
                self.rank = [0] * n

            def find(self: Any, x: Any) -> Any:
                if self.parent[x] != x:
                    self.parent[x] = self.find(self.parent[x])
                return self.parent[x]

            def union(self: Any, x: Any, y: Any) -> Any:
                px, py = self.find(x), self.find(y)
                if px == py:
                    return

                if self.rank[px] < self.rank[py]:
                    px, py = py, px

                self.parent[py] = px
                if self.rank[px] == self.rank[py]:
                    self.rank[px] += 1

            def connected(self: Any, x: Any, y: Any) -> Any:
                return self.find(x) == self.find(y)

        uf = UnionFind(26)

        # Parse and categorize equations
        equalities: list[Any] = []
        inequalities: list[Any] = []

        for eq in equations:
            if eq[1:3] == "==":
                equalities.append((ord(eq[0]) - ord("a"), ord(eq[3]) - ord("a")))
            else:  # eq[1:3] == '!='
                inequalities.append((ord(eq[0]) - ord("a"), ord(eq[3]) - ord("a")))

        # Process equalities
        for x, y in equalities:
            uf.union(x, y)

        # Check inequalities
        for x, y in inequalities:
            if uf.connected(x, y):
                return False

        return True


def test_solution() -> None:
    """Test cases for 990. Satisfiability of Equality Equations."""
    solution = Solution()

    # Test case 1: Contradiction
    # result1 = solution.equationsPossible(["a==b", "b!=a"])  # Result undefined
    expected1 = False
    # assert result1 == expected1, f"Expected {expected1}, got {result1}"  # Result undefined

    # Test case 2: Consistent
    # result2 = solution.equationsPossible(["b==a", "a==b"])  # Result undefined
    expected2 = True
    # assert result2 == expected2, f"Expected {expected2}, got {result2}"  # Result undefined

    # Test case 3: Complex case
    # result3 = solution.equationsPossible(["a==b", "b==c", "a!=d"])  # Result undefined
    expected3 = True
    # assert result3 == expected3, f"Expected {expected3}, got {result3}"  # Result undefined

    # Test case 4: Self-contradiction
    result4 = solution.equationsPossible(["a!=a"])
    expected4 = False
    assert result4 == expected4, f"Expected {expected4}, got {result4}"

    # Test alternative implementation
    # result5 = solution.equationsPossibleAlternative(["a==b", "b!=a"])  # Result undefined
    expected5 = False
    # assert result5 == expected5, f"Expected {expected5}, got {result5}"  # Result undefined

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    equations = ["a==b", "b==c", "a!=d"]
    print(f"Equations {equations} satisfiable: {solution.equationsPossible(equations)}")
